lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/DelegateDecoders.java

```diff
@@ -47,4 +48,31 @@ public void skip(int n)
             }
         };
     }
+
+    public static ValueDecoder<long[]> timeMillisDecoder(ValueDecoder<long[]> delegate)
+    {
+        return new ValueDecoder<>()
+        {
+            @Override
+            public void init(SimpleSliceInputStream input)
+            {
+                delegate.init(input);
+            }
+
+            @Override
+            public void read(long[] values, int offset, int length)
+            {
+                delegate.read(values, offset, length);
+                for (int i = offset; i < offset + length; i++) {
+                    values[i] = values[i] * PICOSECONDS_PER_MILLISECOND;
```

Comment:
Can this overflow or produce value outside of allowed range for TIME Trino type?

Reply:
No such validation is done in either old and new reader for time types. If we read an incorrect parquet file we may end up with incorrect data.

Reply:
The input `values` are decoded from INT32, so multiplying by `PICOSECONDS_PER_MILLISECOND` shouldn't overflow the long. I'm not sure what is the allowed range for TIME Trino type though, I couldn't find a limit from reading the code in TimeType or SqlTime.

Reply:
> 'm not sure what is the allowed range for TIME Trino type though

between 0 and `23:59:59.999999999999` (for precision=12)

Reply:
> I couldn't find a limit from reading the code in TimeType or SqlTime.

https://github.com/trinodb/trino/blob/12816437ec65e45266ec2c24227839ebba6357b7/core/trino-spi/src/main/java/io/trino/spi/type/SqlTime.java#L42-L47

Reply:
Ah right, totally missed that. I was looking for a `writeLong` which wasn't there. I see that `TinyintType#writeLong` for example applies range check but `TimeType` doesn't.
TimeMicrosColumnReader doesn't do anything to check range either.
It seems we're also neglecting rounding for precision < 6  in TimeMicrosColumnReader and precision < 3 here. Maybe we should add checkArgument on TimeType precision as well.